File: 프로그래머스/3/42898. 등굣길/등굣길.py
def solution(m, n, puddles):
    # (1,1) -> (m,n)
    # 오, 아
    dp = [[0] * (m+1) for _ in range(n+1)]
    dp[1][1] = 1
    for y, x in puddles:
        dp[x][y] = -1
    
    # 집에서 학교까지 갈 수 있는 최단경로의 개수
    for i in range(1, n+1):
        for j in range(1, m+1):
            if i == 1 and j == 1:
                continue
            
            if dp[i][j] == -1:
                dp[i][j] = 0
                continue
            
            up = dp[i-1][j]
            left = dp[i][j-1]
            dp[i][j] = (up + left) % 1000000007
    
    return dp[n][m]

File: 프로그래머스/3/42898. 등굣길/등굣길.cpp
#include <string>
#include <vector>

using namespace std;


int solution(int m, int n, vector<vector<int>> puddles) {
    vector<vector<int> > dp(n+1, vector<int>(m+1, 0));
    
    dp[1][1] = 1;
    for (auto p : puddles) {
        dp[p[1]][p[0]] = -1;
    }
    
    for (int i = 1; i < n+1; i++) {
        for (int j = 1; j < m+1; j++) {
            
            if (i==1 && j==1) continue;
            if (dp[i][j] == -1) {
                dp[i][j] = 0;
                continue;
            }
            
            dp[i][j] = (dp[i-1][j] + dp[i][j-1]) % 1000000007;
        }
    }
    
    return dp[n][m];
}